vim-patch:8.1.2073: when editing a buffer 'colorcolumn' may not work

zeertzjq opened the pull request:
#### vim-patch:8.1.2073: when editing a buffer 'colorcolumn' may not work

Problem:    When editing a buffer 'colorcolumn' may not work.
Solution:   Set the buffer before copying option values. Call
            check_colorcolumn() after copying window options.
https://github.com/vim/vim/commit/010ee9657acf1a9f799079d718998c94e50ccadc

Fix #12918

zeertzjq commented:
The test I added fails on `master`:
```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] test/functional/ui/highlight_spec.lua @ 1203: ColorColumn highlight when entering a buffer vim-patch:8.1.2073
test/functional/ui/highlight_spec.lua:1214: Row 8 did not match.
Expected:
  |{4:  1 }11{1:1}11111{1:1}1                          |
  |{3:  2 }22{1:2}22222{1:2}22                         |
  |{3:  3 }33{1:3}33333{1:3}3                          |
  |{5:~                                       }|
  |{5:~                                       }|
  |{5:~                                       }|
  |{7:X                                       }|
  |*{4:  1 }^11{1:1}11111{1:1}1                          |
  |*{3:  2 }22{1:2}22222{1:2}22                         |
  |*{3:  3 }33{1:3}33333{1:3}3                          |
  |{5:~                                       }|
  |{5:~                                       }|
  |{5:~                                       }|
  |{6:X                                       }|
  |                                        |
Actual:
  |{4:  1 }11{1:1}11111{1:1}1                          |
  |{3:  2 }22{1:2}22222{1:2}22                         |
  |{3:  3 }33{1:3}33333{1:3}3                          |
  |{5:~                                       }|
  |{5:~                                       }|
  |{5:~                                       }|
  |{7:X                                       }|
  |*{4:  1 }^1111111111                          |
  |*{3:  2 }22222222222                         |
  |*{3:  3 }3333333333                          |
  |{5:~                                       }|
  |{5:~                                       }|
  |{5:~                                       }|
  |{6:X                                       }|
  |                                        |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  

stack traceback:
	test/functional/ui/screen.lua:595: in function '_wait'
	test/functional/ui/screen.lua:362: in function 'expect'
	test/functional/ui/highlight_spec.lua:1214: in function <test/functional/ui/highlight_spec.lua:1203>


 1 FAILED TEST
```

seandewar commented on src/nvim/option.c:
```diff
@@ -6136,6 +6135,7 @@ void win_copy_options(win_T *wp_from, win_T *wp_to)
 {
   copy_winopt(&wp_from->w_onebuf_opt, &wp_to->w_onebuf_opt);
   copy_winopt(&wp_from->w_allbuf_opt, &wp_to->w_allbuf_opt);
+  didset_window_options(wp_to);
```
Nvim already does this in win_init.

zeertzjq replied:
Thanks!

zeertzjq committed:
test: convert some colorcolumn tests to Lua screen tests

zeertzjq committed:
vim-patch:8.1.2073: when editing a buffer 'colorcolumn' may not work

Problem:    When editing a buffer 'colorcolumn' may not work.
Solution:   Set the buffer before copying option values. Call
            check_colorcolumn() after copying window options.
https://github.com/vim/vim/commit/010ee9657acf1a9f799079d718998c94e50ccadc
